platform/ui/src/components/userPreferences/HotkeysPreferences.js

```diff
@@ -0,0 +1,183 @@
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+
+import { TabFooter } from './TabFooter';
+import { HotkeyField } from '../customForm';
+import { hotkeysValidators } from './hotkeysValidators';
+import { MODIFIER_KEYS, ALLOWED_KEYS } from './hotkeysConfig';
+
+import { useSnackbarContext } from '@ohif/ui';
+
+import './HotkeysPreferences.styl';
+
+const initialState = hotkeyDefinitions => ({
+  hotkeys: { ...hotkeyDefinitions },
+  errors: {},
+});
+
+const validateCommandKey = ({ commandName, pressedKeys, hotkeys }) => {
+  for (const validator of hotkeysValidators) {
+    const validation = validator({
+      commandName,
+      pressedKeys,
+      hotkeys,
+    });
+    if (validation && validation.hasError) {
+      return validation;
+    }
+  }
+
+  return {
+    hasError: false,
+    errorMessage: undefined,
+  };
+};
+
+/**
+ * HotkeysPreferences tab
+ * It renders all hotkeys displayed into columns/rows
+ *
+ * It stores current state and whenever it changes, component messages parent of new value (through function callback)
+ * @param {object} props component props
+ * @param {string} props.onClose
+ * @param {object} props.t
+ * @param {object} porps.hotkeyDefinitions
+ * @param {object} porps.hotkeyDefaults
+ * @param {object} porps.setHotkeys
+ */
+function HotkeysPreferences({
+  onClose,
+  t,
+  hotkeyDefinitions,
+  hotkeyDefaults,
+  setHotkeys,
+}) {
+  const [state, setState] = useState(initialState(hotkeyDefinitions));
+
+  const snackbar = useSnackbarContext();
+
+  const onResetPreferences = () => {
+    const defaultHotKeyDefinitions = {};
+
+    hotkeyDefaults.map(item => {
+      const { commandName, ...values } = item;
+      defaultHotKeyDefinitions[commandName] = { ...values };
+    });
+
+    setState(initialState(defaultHotKeyDefinitions));
+  };
+
+  const onSave = () => {
+    const { hotkeys } = state;
+
+    setHotkeys(hotkeys);
+
+    localStorage.setItem('hotkey-definitions', JSON.stringify(hotkeys));
+
+    onClose();
+
+    snackbar.show({
+      message: t('SaveMessage'),
+      type: 'success',
+    });
+  };
+
+  const onHotkeyChanged = (commandName, hotkeyDefinition, keys) => {
+    const { errorMessage } = validateCommandKey({
+      commandName,
+      pressedKeys: keys,
+      hotkeys: state.hotkeys,
+    });
+
+    setState(prevState => ({
+      hotkeys: {
+        ...prevState.hotkeys,
+        [commandName]: { ...hotkeyDefinition, keys },
+      },
+      errors: {
+        ...prevState.errors,
+        [commandName]: errorMessage,
+      },
+    }));
+  };
+
+  const hasErrors = Object.keys(state.errors).some(key => !!state.errors[key]);
+  const hasHotkeys = Object.keys(state.hotkeys).length;
+
+  const splitedHotkeys = [];
+  if (hasHotkeys) {
+    const arrayHotkeys = Object.entries(state.hotkeys);
+    const halfwayThrough = Math.ceil(arrayHotkeys.length / 2);
+    splitedHotkeys.push(arrayHotkeys.slice(0, halfwayThrough));
```

Comment:
For blocks of code like this one which does some complicated stuff like entries -> ceil -> slice -> pushing, it's a good idea to extract into another well-named function which documents what it's doing. This helps a lot reading it in the future =)

Reply:
Change done.